### PR TITLE
feat(ci): add static CI health dashboard for GitHub Pages

### PR DESCRIPTION
**Pull request by kami619**
## Description
  Adds a lightweight, read-only CI health dashboard for RHAIENG-5940 (QG8 stretch goal under RHAIENG-5841).
  This PR introduces:
  - A Python generator (`.github/scripts/generate_ci_health_page.py`) that summarizes the four in-scope QG8 workflows from the GitHub Actions API or a committed fixture
  - A `CI Health Pages` workflow that builds static HTML and deploys it to GitHub Pages via `configure-pages` / `upload-pages-artifact` / `deploy-pages`
  - Documentation for data source, update cadence, and bootstrap steps (`docs/ci-health-dashboard.md`)
  The dashboard focuses on shared-branch health (`main` push, `schedule`, and `workflow_dispatch` runs). PR runs are excluded to avoid noise.
  **Design decisions:**
  - Static HTML only (no MkDocs) for this spike; runbooks can be proposed separately based on feedback
  - `workflow_dispatch` supports a `use_fixture` input for safe pre-merge demos
  - Daily refresh at 06:00 UTC (after the QG4 nightly window)
  ## Jira Ticket
  RHAIENG-5940
  ## Testing
  - [ ] `make test` passes (run from the affected agent directory)
  - [x] Manual testing performed (describe steps below)
  - [ ] No testing required (documentation/config change only)
  - `python -m pytest .github/scripts/tests/test_generate_ci_health_page.py -v` — 4 tests passed
  - `ruff check` on generator and test files — passed
  - Generated sample page locally with `--input .github/scripts/fixtures/ci-runs-sample.json`
  **Post-merge / PR-branch verification:**
  1. Actions → **CI Health Pages** → Run workflow on this branch with `use_fixture: true`
  2. Confirm deploy job succeeds and Pages URL serves the dashboard
  3. Re-run with live API data after merge
  ## Checklist
  - [x] I have read [CONTRIBUTING.md](CONTRIBUTING.md)
  - [x] No `.env` or secret files are included in this PR
  - [x] All changes are within scope of the linked Jira ticket (if not, explain in Description)
  ## Review Guidance
  - `.github/scripts/generate_ci_health_page.py` — core logic and HTML output
  - `.github/workflows/ci-health-pages.yml` — Pages permissions, fixture vs live generation, schedule
  - `docs/ci-health-dashboard.md` — operator setup (Pages must be enabled with Source: GitHub Actions)
  ## Related PRs
  N/A (first CI health dashboard PR for QG8)

### CODE REVIEW

File: .github/scripts/tests/test_generate_ci_health_page.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            html_url="https://example.com/2",
+            created_at="2026-07-05T11:00:00Z",
+            updated_at="2026-07-05T11:05:00Z",
+            run_started_at="2026-07-05T11:00:10Z",
+        ),
+    ]
+    rate, total, passed = compute_pass_rate(runs, days=7)
+    assert total == 1
+    assert passed == 1
+    assert rate == 100.0
```
[the remark this answers is not in the document]
Comment (kami619) [reply .github/scripts/tests/test_generate_ci_health_page.py:70]: fixed in https://github.com/red-hat-data-services/agentic-starter-kits/pull/234/commits/ce314f4220c9cba108c58ae1085c622b6f64bba2

File: .github/scripts/generate_ci_health_page.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+            with urllib.request.urlopen(request, timeout=30) as response:
+                return json.load(response)
+        except urllib.error.HTTPError as exc:
+            body = exc.read().decode("utf-8", errors="replace")
+            raise RuntimeError(f"GitHub API error {exc.code} for {path}: {body}") from exc
+
+    def list_workflow_runs(self, workflow_file: str, *, per_page: int = 30) -> list[WorkflowRun]:
+        path = f"/repos/{self.repository}/actions/workflows/{workflow_file}/runs"
+        payload = self._request(path, {"per_page": per_page})
+        return [WorkflowRun.from_api(item) for item in payload.get("workflow_runs", [])]
```
[the remark this answers is not in the document]
Comment (kami619) [reply .github/scripts/generate_ci_health_page.py]: fixed in https://github.com/red-hat-data-services/agentic-starter-kits/pull/234/commits/ce314f4220c9cba108c58ae1085c622b6f64bba2

File: .github/scripts/generate_ci_health_page.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+        url = f"{API_ROOT}{path}"
+        if query:
+            url = f"{url}?{urllib.parse.urlencode(query)}"
+        request = urllib.request.Request(url, headers=self.headers)
+        try:
+            with urllib.request.urlopen(request, timeout=30) as response:
+                return json.load(response)
+        except urllib.error.HTTPError as exc:
+            body = exc.read().decode("utf-8", errors="replace")
+            raise RuntimeError(f"GitHub API error {exc.code} for {path}: {body}") from exc
```
[the remark this answers is not in the document]
Comment (kami619) [reply .github/scripts/generate_ci_health_page.py]: fixed in https://github.com/red-hat-data-services/agentic-starter-kits/pull/234/commits/ce314f4220c9cba108c58ae1085c622b6f64bba2

File: .github/scripts/generate_ci_health_page.py [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+        return f"{minutes}m {secs}s"
+    return f"{secs}s"
+
+
+def is_relevant_run(run: WorkflowRun) -> bool:
+    if run.event not in RELEVANT_EVENTS:
+        return False
+    if run.event == "push" and run.head_branch != "main":
+        return False
+    return True
```
[the remark this answers is not in the document]
Comment (kami619) [reply .github/scripts/generate_ci_health_page.py]: fixed in https://github.com/red-hat-data-services/agentic-starter-kits/pull/234/commits/ce314f4220c9cba108c58ae1085c622b6f64bba2

**Comment by kami619** [comment]
@shricharan-ks take a look at this PR and see if you can merge it, I have enabled the GH pages for the repo level, once its merged we can test the GH pages rendering of the CI run results at https://red-hat-data-services.github.io/agentic-starter-kits/

**Review by shricharan-ks — APPROVED** [review]
## Code Review — PR #234

**Scope**: New static CI health dashboard for GitHub Pages. 6 new files, +828 lines (all additions). Python generator script + GitHub Actions workflow + tests + fixture + docs. All CI green.

**Verdict**: APPROVE — well-structured, clean implementation with good security practices. Minor suggestions below.

---

### Architecture

Clean separation: data fetching (API/fixture) → summarization (`WorkflowSummary` dataclass) → HTML rendering. Frozen dataclasses, `from_api` classmethod with fallbacks for missing API fields, and dual-mode operation (fixture for offline/demos, live for production). The script has zero external dependencies — pure stdlib (`urllib`, `json`, `html`, `argparse`, `dataclasses`).

---

### Issues

#### 1. Fixture dates will age out of the 7-day pass-rate window (Minor)

`ci-runs-sample.json` uses hardcoded dates like `2026-07-05`. The `compute_pass_rate()` function computes a 7-day cutoff from `datetime.now(UTC)`. After ~July 12, fixture-mode dashboards will show "No completed runs in the last 7 days" for all workflows.

This is acceptable since fixture mode is for demos and local testing only, not production — but worth noting in the doc or as a comment in the fixture file, so future contributors don't wonder why the fixture dashboard looks empty.

#### 2. Test coverage gaps — no tests for edge cases (Minor)

The 5 tests cover the critical paths well. Missing edge cases:

- **Empty workflow** (no runs at all) — does `summarize_workflow()` handle it? (It does — `latest = None`, `pass_rate = None` — but no test verifies this)
- **`schedule` events included regardless of branch** — `is_relevant_run()` has special handling for schedule but no test for it
- **`duration_seconds()` / `format_duration()`** — untested utility functions
- **`conclusion_label()` mapping** — untested
- **Pass rate with mixed results** (e.g., 2 pass + 1 fail = 66.7%) — only 100% case tested

Not blocking — the existing tests verify the most important invariants (PR filtering, feature-branch exclusion, end-to-end HTML generation).

#### 3. `sys.path.insert` import mechanism (Nit)

`test_generate_ci_health_page.py` uses `sys.path.insert(0, str(Path(__file__).resolve().parent.parent))` to import the generator. This is fragile and IDE-unfriendly. An alternative would be adding a minimal `conftest.py` that does the path fixup, or using `importlib.import_module`. Not blocking — this pattern is common for CI scripts that aren't part of a proper Python package.

---

### Positive Observations

- **HTML escaping is thorough** — all API-sourced strings go through `html.escape()`, URLs use `html.escape(url, quote=True)` for attribute context. No XSS vectors.
- **GitHub Actions workflow security**:
  - All 5 actions pinned by SHA commit hash (checkout, setup-python, configure-pages, upload-pages-artifact, deploy-pages)
  - `persist-credentials: false` on checkout prevents token leakage
  - `timeout-minutes: 10` prevents runaway jobs
  - `if: github.repository == '...'` guard prevents fork deployments
- **Permissions are minimal**: `contents: read`, `pages: write`, `id-token: write` (OIDC for Pages), `actions: read` (list workflow runs). Nothing extra.
- **`is_relevant_run()` logic is correct**:
  - `schedule` → always included (nightly runs regardless of branch)
  - `push` to `main` → included; to feature branches → excluded
  - `workflow_dispatch` on `main` → included; on feature branches → excluded
  - `pull_request` → excluded (not in `RELEVANT_EVENTS`)
- **Pass rate calculation handles edge cases**: `cancelled` and `skipped` runs excluded from denominator, `timed_out` counted as failure, empty result set returns `None` instead of division-by-zero.
- **API pagination is bounded**: max 10 pages × 100 runs = 1000 runs max, with early exit on cutoff or short page. No infinite loop risk.
- **Dark/light mode CSS** via `prefers-color-scheme` media query — nice touch.
- **Concurrency** `cancel-in-progress: false` is correct for Pages deployments (don't cancel mid-deploy).
- **Documentation** is thorough — covers scope, setup (including pre-merge Pages enablement), local preview with both modes, and test commands.

---

### Summary

This is a clean, well-tested dashboard feature. The Python script is well-structured with proper HTML escaping, the workflow has excellent security practices, and the documentation makes onboarding easy. The fixture-mode aging and test coverage gaps are minor — nothing blocks merging and enabling GitHub Pages.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

**Comment by kami619** [comment]
@shricharan-ks I ack the nits, merging it move the testing further. we can harden as we add more features. Thanks again for your review.

**Comment by tarun-etikala** [comment]
## Feedback

### 1. No resilience per workflow API call

If one workflow API call fails (e.g., `eval-gating.yml` 404s because it was renamed), the entire generation crashes and no dashboard gets published. Consider catching per-workflow errors and rendering a "data unavailable" card instead of failing the whole job.

### 2. `workflow_dispatch` from feature branches overwrites production Pages

There's no branch guard on the `workflow_dispatch` trigger — someone running it from a non-main branch would overwrite the production GitHub Pages site with potentially stale or incorrect data. Consider adding a condition like:

```yaml
if: github.ref == 'refs/heads/main' || (github.event_name == 'workflow_dispatch' && inputs.use_fixture == true)
```

Or at minimum restrict the deploy step to `main` so that feature-branch dispatches generate the artifact without publishing.
